File: Python/19. Remove Nth Node From End of List.py
from ListNode import *

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        newList = ListNode()
        newList.next = head
        nodes = 0
        first = head
        
        while first:
            nodes+=1
            first = first.next
            
        nodes-=n
        first = newList
        
        while nodes>0:
            nodes-=1
            first = first.next
            
        first.next = first.next.next
        return newList.next
    
if __name__ == "__main__":
    head = ListNode(1,ListNode(2,ListNode(3,ListNode(4,ListNode(5)))))
    n = 2
    result = Solution().removeNthFromEnd(head,n)
    print(result)

File: CPP/19. Remove Nth Node From End of List.cpp
#include <iostream>
#include "printFunctions.hpp"

using namespace std;

class Solution {
public:
	ListNode* removeNthFromEnd(ListNode* head, int n) {
		ListNode* newList = new ListNode(0);
		newList->next = head;
		ListNode* first = head;
		int nodes = 0;

		while (first) { // while (first != nullptr) { is the same
			nodes++;
			first = first->next;
		}

		nodes -= n;
		first = newList;

		while (nodes > 1) {
			nodes--;
			first = first->next;
		}

		first->next = first->next->next;

		return newList->next;
	}
};


int main() {
	Solution sol;
	ListNode* head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
	int n = 2;
	ListNode* result = sol.removeNthFromEnd(head,n);
	printListNode(result);
	return 0;
}